src/transformers/models/bark/modeling_bark.py

```diff
@@ -0,0 +1,1372 @@
+                ),
+                "wpe": nn.Embedding(config.block_size, config.hidden_size),
+                "drop": nn.Dropout(config.dropout),
+                "h": nn.ModuleList(
+                    [BarkBlock(config, layer_idx, is_causal=False) for layer_idx in range(config.num_layers)]
+                ),
+                "ln_f": nn.LayerNorm(config.hidden_size),
+            }
+        )
+        self.lm_heads = nn.ModuleList(
+            [
+                nn.Linear(config.hidden_size, config.output_vocab_size, bias=False)
+                for _ in range(config.n_codes_given, config.n_codes_total)
+            ]
+        )
+        for i in range(config.n_codes_total - config.n_codes_given):
+            self.transformer.wtes[i + 1].weight = self.lm_heads[i].weight
+
+    def get_num_params(self, non_embedding=True):
+        """
+        Return the number of parameters in the model. For non-embedding count (default), the position embeddings get
+        subtracted. The token embeddings would too, except due to the parameter sharing these params are actually used
+        as weights in the final layer, so we include them.
+        """
+        n_params = sum(p.numel() for p in self.parameters())
+        if non_embedding:
+            for wte in self.transformer.wtes:
+                n_params -= wte.weight.numel()
+            n_params -= self.transformer.wpe.weight.numel()
+        return n_params
+
+    def _get_and_check_input_embeddings(self, input_ids, input_embeds, pred_idx):
+        # the input_embeddings are the sum of the j previous codebooks embeddings before the current pred_idx codebook
+
+        if input_ids is not None and input_embeds is not None:
+            raise ValueError("You cannot specify both input_ids and input_embeds at the same time")
+        elif input_ids is not None:
+            batch_size, seq_len, codes = input_ids.size()
+            assert (
+                batch_size <= self.config.block_size
+            ), f"Cannot forward sequence of length {seq_len}, block size is only {self.config.block_size}"
+
+            assert codes == self.n_codes_total, (batch_size, seq_len, codes)
+
+            # forward the GPT model itself
+            input_embeds = [
+                wte(input_ids[:, :, i]).unsqueeze(-1) for i, wte in enumerate(self.transformer.wtes)
+            ]  # token embeddings of shape (b, t, n_embd)
+            input_embeds = torch.cat(input_embeds, dim=-1)
+            input_embeds = input_embeds[:, :, :, : pred_idx + 1].sum(dim=-1)
+
+        elif input_embeds is not None:
+            input_embeds = self.transformer.wte(input_ids)  # token embeddings of shape (b, t, n_embd)
+        else:
+            raise ValueError("You have to specify either input_ids or input_embeds")
+
+        return input_embeds
+
+    # contrary to the other main module of Bark (BarkCausalModule), it is non-causal, so no need for past key values
```

Comment:
This would be very clear if we defined the base model above, and then initialised it here with `causal=False` and the LM heads